### PR TITLE
JIT: refactor instrumentation code

### PR DESCRIPTION
**Pull request by AndyAyersMS**
Refactor `fgInstrument` and related methods and classes, so that the strategies
for block and class instrumentation are clearly separated out, and can be
varied.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@dotnet/jit-contrib PTAL
cc @davidwrighton 

This is essentially a no-diff change, save that the schemas will now interleave count and class probe elements. If that's undesirable I can revise to keep them in separate ranges of the schema. I am sharing some under-utilized fields in BasicBlock to keep track of the block to (base) schema index.

This should be the last prerequisite change before I implement the sparse instrumentation scheme described in #46882.

Might be a bit tricky to review given all the code that has moved around.

**Comment by AndyAyersMS** [comment]
Experimental failures are the "expected" ones in OSR/EhWriteThru (same as in #47476). All the PGO tests passed.

File: src/coreclr/jit/fgprofile.cpp [hunk 1]
```diff
@@ -244,387 +465,407 @@ class ClassProbeVisitor final : public GenTreeVisitor<ClassProbeVisitor<TFunctor
...
+        visitor.WalkTree(stmt->GetRootNodePointer(), nullptr);
+    }
+}
 
-    for (block = fgFirstBB; (block != nullptr); block = block->bbNext)
+//------------------------------------------------------------------------
+// ClassProbeInstrumentor::SuppressProbes: clean up if we're not instrumenting
+//
+// Notes:
+//   Currently we're hijacking the gtCallStubAddre of the call node to hold
```
Comment (BruceForstall) [inline src/coreclr/jit/fgprofile.cpp:710]: nit: => gtStubCallStubAddre => gtStubCallStubAddr

**Review by BruceForstall — APPROVED** [review]
LGTM

File: src/coreclr/jit/fgprofile.cpp (hunk 1 above)
Comment (AndyAyersMS) [reply src/coreclr/jit/fgprofile.cpp:710]: Thanks. Will fix in the next PR...

**Pull request merged**